src: generate snapshot with --predictable

joyeecheung opened the pull request:
To improve determinism of snapshot generation, add --predictable to the V8 flags used to initialize a process launched to generate snapshot. Also add a kGeneratePredictableSnapshot flag to ProcessInitializationFlags for this and moves the configuration of these flags into node::InitializeOncePerProcess() so that it can be shared by embedders.

Refs: https://github.com/nodejs/build/issues/3043

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

anonrig commented on src/node.cc:
```diff
@@ -932,6 +932,13 @@ InitializeOncePerProcessInternal(const std::vector<std::string>& args,
+  if (!!(flags & ProcessInitializationFlags::kGeneratePredictableSnapshot) ||
+      per_process::cli_options->per_isolate->build_snapshot) {
+    v8::V8::SetFlagsFromString("--predictable");
+    v8::V8::SetFlagsFromString("--random_seed=42");
```
Is there a specific reason for 42? I know 42 is the meaning of life, the universe, and everything (Referencing Hitchikers guide to the galaxy)

joyeecheung replied:
It was added by https://github.com/nodejs/node/pull/29142, probably just picked randomly

bnoordhuis replied:
Madam, I will have you know I have [a hobby of sneaking the number 42 into the node.js code base](
https://github.com/nodejs/node/commit/b4def4806c50b5aa6a71376c41bb1d5cbe41d29a#diff-25cedd03572aede0ca494eea62d581431b58ef9da9dc536ec64fc25810655cfbR19) for almost 13 years now. I plan to keep it up for another 29.

joyeecheung commented:
This needs https://chromium-review.googlesource.com/c/v8/v8/+/4681766

targos commented:
https://chromium-review.googlesource.com/c/v8/v8/+/4681766 landed.

joyeecheung commented:
It turns out all the flags implied by --predictable need to be ignored in the flag hash computation too:

```
--no-concurrent-recompilation
--no-parallel-scavenge
--no-concurrent-marking
--no-concurrent-array-buffer-sweeping
--no-parallel-marking
--no-concurrent-sweeping
--no-parallel-compaction
--no-parallel-pointer-update
--no-memory-reducer
--single-threaded-gc
```

Although I now wonder if `--predictable` is actually needed for our current use case (it seems `--predictable` currently does not actually introduction additional changes to the snapshot?) This might still be needed down the line though, I may come back and pick this up later when it turns out to be necessary

joyeecheung commented:
hmm, looks like this is necessary after all or otherwise we get 2 more differences: https://github.com/nodejs/build/issues/3043#issuecomment-1692117367

joyeecheung commented:
Rebased after https://github.com/nodejs/node/pull/49703 is backported and also moved the flag settings into `InitializeNodeWithArgsInternal()`, together with other `V8::SetFlagsFromString` calls.

joyeecheung commented:
@anonrig @richardlau This has been rebased and updated. Can I get some new LGTM please?
